lib/datadog/appsec/context.rb

```diff
@@ -34,18 +32,33 @@ def active
       def initialize(trace, span, security_engine)
         @trace = trace
         @span = span
+        @events = []
         @security_engine = security_engine
-        @waf_runner = security_engine.new_context
+        @waf_runner = security_engine.new_runner
+        @waf_metrics = WAFMetrics.new(timeouts: 0, duration_ns: 0, duration_ext_ns: 0)
+        @mutex = Mutex.new
       end
 
       def run_waf(persistent_data, ephemeral_data, timeout = WAF::LibDDWAF::DDWAF_RUN_TIMEOUT)
-        @waf_runner.run(persistent_data, ephemeral_data, timeout)
+        result = @waf_runner.run(persistent_data, ephemeral_data, timeout)
+
+        @mutex.synchronize do
+          @waf_metrics.timeouts += 1 if result.timeout?
+          @waf_metrics.duration_ns += result.duration_ns
+          @waf_metrics.duration_ext_ns += result.duration_ext_ns
+        end
```

Comment:
what do you think, should we move the mutex inside `WAFMetrics`?

Reply:
Yes, I have that in mind, I think for that we need to extract the `Metrics` tool which will do that because `WAFMetrics` is a value-object.